include/boost/math/special_functions/interval_sieve.hpp

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 Matt Borland and Jonathan Sorenson
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_INTERVAL_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_INTERVAL_SIEVE_HPP
+
+#include <boost/dynamic_bitset.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
+#include <boost/math/special_functions/prime_wheel.hpp>
+#include <cmath>
+#include <memory>
+#include <array>
+#include <cstdint>
+#include <bitset>
+
+namespace boost::math::detail
+{
+
+#if defined(__MPIR_VERSION) || defined(__GNU_MP_VERSION)
+// GNU GMP C or MPIR
+inline double get_double(const mpz_t &x) noexcept
+{
+    return mpz_get_d(x);
+}
+#endif
+
+#if defined(__GNU_MP_VERSION)
+#if __has_include(<gmpxx.h>)
+// GNU GMP C++ bindings
+inline double get_double(const mpz_class &x) noexcept
+{
+    return x.get_d()
+}
+#endif
+#endif
+
+// boost::multiprecision and POD
+template<class Integer>
+inline double get_double(const Integer &x) noexcept
+{
+    return static_cast<double>(x);
+}
+
+template<class Integer, class PrimeContainer, class Container>
+class IntervalSieve
+{
+#ifdef __SIZEOF_INT128__   // Defined in GCC 4.6+, clang, intel. MSVC does not define. 
```

Comment:
Better to use BOOST_HAS_INT128 as there are some corner cases where __SIZEOF_INT128__  does the wrong thing.